cardano-api/src/Cardano/Api/Eras.hs

```diff
@@ -365,6 +383,62 @@ instance IsShelleyBasedEra BabbageEra where
 instance IsShelleyBasedEra ConwayEra where
    shelleyBasedEra = ShelleyBasedEraConway
 
+data AnyShelleyBasedEra where
+  AnyShelleyBasedEra
+    :: IsShelleyBasedEra era  -- Provide class constraint
+    => ShelleyBasedEra era    -- and explicit value.
+    -> AnyShelleyBasedEra
+
+deriving instance Show AnyShelleyBasedEra
+
+instance Eq AnyShelleyBasedEra where
+    AnyShelleyBasedEra era == AnyShelleyBasedEra era' =
+      case testEquality era era' of
+        Nothing   -> False
+        Just Refl -> True -- since no constructors share types
+
+instance Bounded AnyShelleyBasedEra where
+   minBound = AnyShelleyBasedEra ShelleyBasedEraShelley
+   maxBound = AnyShelleyBasedEra ShelleyBasedEraConway
+
+instance Enum AnyShelleyBasedEra where
+   enumFrom e = enumFromTo e maxBound
+
+   fromEnum = \case
+      AnyShelleyBasedEra ShelleyBasedEraShelley  -> 1
```

Comment:
Why manual instance, instead of deriving one for this type and `ShelleyBasedEra`?

Reply:
This is what happens when attempting this:

```
src/Cardano/Api/Eras.hs:394:1: error:
    • Couldn't match type ‘era1’ with ‘era’
      ‘era1’ is a rigid type variable bound by
        a pattern with constructor:
          AnyShelleyBasedEra :: forall era.
                                IsShelleyBasedEra era =>
                                ShelleyBasedEra era -> AnyShelleyBasedEra,
        in an equation for ‘==’
        at src/Cardano/Api/Eras.hs:394:1-39
      ‘era’ is a rigid type variable bound by
        a pattern with constructor:
          AnyShelleyBasedEra :: forall era.
                                IsShelleyBasedEra era =>
                                ShelleyBasedEra era -> AnyShelleyBasedEra,
        in an equation for ‘==’
        at src/Cardano/Api/Eras.hs:394:1-39
      Expected type: ShelleyBasedEra era
        Actual type: ShelleyBasedEra era1
    • In the second argument of ‘(==)’, namely ‘b1’
      In the expression: ((a1 == b1))
      In an equation for ‘==’:
          (==) (AnyShelleyBasedEra a1) (AnyShelleyBasedEra b1) = ((a1 == b1))
      When typechecking the code for ‘==’
        in a derived instance for ‘Eq AnyShelleyBasedEra’:
        To see the code I am typechecking, use -ddump-deriv
    • Relevant bindings include
        b1 :: ShelleyBasedEra era1 (bound at src/Cardano/Api/Eras.hs:394:1)
        a1 :: ShelleyBasedEra era (bound at src/Cardano/Api/Eras.hs:394:1)
    |
394 | deriving instance Eq AnyShelleyBasedEra
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Thanks for checking that. I see that we can't have derived classes with existential quantification in constructor.